pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,98 @@
+package cluster
+
+import (
+	"encoding/json"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GetClusterAgentTolerations returns additional tolerations for the cluster agent if they have been user defined. If
+// not, nil is returned.
+func GetClusterAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetClusterAgentAffinity returns node affinity for the cluster agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	affinity, err := unmarshalAffinity(settings.ClusterAgentDefaultAffinity.Get())
+	if err != nil {
```

Comment:
We shouldn't discard this error.

Let's draw a hypothetical situation -- there is a cluster where the affinity is required is allow the cluster-agent to connect to Rancher, for example, there are only a few nodes within the cluster that can connect to the network where Rancher lives. We use the `cluster-deploy` controller to manage the `cluster-agent` in the downstream cluster, and an admin is in the process of trying to make a change to the `ClusterAgentDefaultAffinity` setting, and makes an error that causes the affinity to not unmarshal (for example, extra punctuation). With the logic written as defined, it is possible that the cluster-deploy controller picks up the broken affinity setting and "falls back" to using Rancher's default cluster-agent affinity -- it delivers a new cluster-agent manifest that proceeds to disconnect the downstream cluster from Rancher because the affinity settings don't steer the cluster-agent to a desired node.

Due to the "reverse-proxy" nature of the cluster-agent, recovery from this situation is not necessarily easy, especially if the downstream cluster API server is not easily accessible. 

This is not ideal and can happen to any of the downstream clusters -- effectively causing a sev1 outage.